### ISSUE TITLE
Kubelet on Fedora 31+ doesn't start without kernel parameter systemd.unified_cgroup_hierarchy = 0

### ISSUE THREAD

**Issue by Vascko** [kubernetes/kubernetes#90710]
**What happened**:
When deploying kubernetes via kubeadm on Fedora 31+ with cri-o, kubelet fails to start with the following error:
`kubelet.go:1383] Failed to start ContainerManager system validation failed - Following Cgroup subsystem not mounted: [cpu cpuacct cpuset memory]`

Subsequently kubeadm can't finish the deployment.

```
I0503 19:08:32.558943   11492 waitcontrolplane.go:87] [wait-control-plane] Waiting for the API server to be healthy
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[kubelet-check] Initial timeout of 40s passed.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get http://localhost:10248/healthz: dial tcp [::1]:10248: connect: connection refused.
```

Fedora 31 has cgroups v2 enabled by default and while CRI-O supports it, I'm not sure if kubernetes does. I could not find a lot of documentation around kubernetes and cgroup v2.

**What you expected to happen**:
kubelet starting and kubeadm finishing the deployment.

**How to reproduce it (as minimally and precisely as possible)**:
run `kubeadm init` on any default Fedora 31+ environment

**Anything else we need to know?**:
When appending `systemd.unified_cgroup_hierarchy = 0` to the kernel command line, the service starts but kubeadm still can't finish the deployment. This time it doesn't report the connection refused error:
```
I0503 19:26:41.473016     680 waitcontrolplane.go:87] [wait-control-plane] Waiting for the API server to be healthy
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[kubelet-check] Initial timeout of 40s passed.

	Unfortunately, an error has occurred:
		timed out waiting for the condition

	This error is likely caused by:
		- The kubelet is not running
```

There is a ton of systemd errors in the kubelet log but nothing obvious (to me) that would keep kubeadm from deploying properly:
```
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246039     797 factory.go:177] Factory "systemd" was unable to handle container "/system.slice/system-systemd\\x2dbacklight.slice"
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246050     797 factory.go:170] Factory "raw" can handle container "/system.slice/system-systemd\\x2dbacklight.slice", but ignoring.
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246064     797 manager.go:908] ignoring container "/system.slice/system-systemd\\x2dbacklight.slice"
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246072     797 factory.go:177] Factory "crio" was unable to handle container "/system.slice/crond.service"
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246083     797 factory.go:166] Error trying to work out if we can handle /system.slice/crond.service: /system.slice/crond.service not handled by systemd handler
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246091     797 factory.go:177] Factory "systemd" was unable to handle container "/system.slice/crond.service"
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246100     797 factory.go:170] Factory "raw" can handle container "/system.slice/crond.service", but ignoring.
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246111     797 manager.go:908] ignoring container "/system.slice/crond.service"
May 03 19:40:46 localhost.localdomain kubelet[797]: I0503 19:40:46.246119     797 factory.go:177] Factory "crio" was unable to handle container "/system.slice/systemd-tmpfiles-setup-dev.service"
kubelet.go:2184] Container runtime status: Runtime Conditions: RuntimeReady=true reason: message:, NetworkReady=true reason: message:
handler.go:175] error while reading "/proc/797/fd/5" link: readlink /proc/797/fd/5: no such file or directory

```

**Environment**:
- Kubernetes version (use `kubectl version`):
`kubeadm version: &version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:54:15Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}`
- Cloud provider or hardware configuration: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz with 16GB of memory
- OS (e.g: `cat /etc/os-release`): Fedora 31+
- Kernel (e.g. `uname -a`): `Linux localhost.localdomain 5.6.8-200.fc31.x86_64 #1 SMP Wed Apr 29 19:10:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: cri-o v1.18.0

**Comment by neolit123** [comment]
> Fedora 31 has cgroups v2 enabled by default and while CRI-O supports it, I'm not sure if kubernetes does. I could not find a lot of documentation around kubernetes and cgroup v2.

the plan is to enable initial cgroups v2 support in 1.19.

the proposal and initial related PRs have merged:
https://github.com/kubernetes/enhancements/pull/1370
https://github.com/kubernetes/kubernetes/pull/89897
https://github.com/kubernetes/kubernetes/pull/85218
https://github.com/kubernetes/kubernetes/pull/89901
(yet unclear what else has to be done)

if you want to help testing download the kubelet and kubeadm binaries from e.g. this CI build and please report if you see issues:
https://console.cloud.google.com/storage/browser/kubernetes-release-dev/ci/v1.19.0-alpha.2.349+de66c643c479f4-bazel/bin/linux/amd64/

/sig cluster-lifecycle
/area kubeadm kubelet

**Comment by neolit123** [comment]
but note that kubelet being newer than the control-plane is not supported, so you need to pass `kubernetesVersion: k8s-master` or `--kubernetes-version=k8s-master` to kubeadm, which should download the latest control-plane images.

**Comment by Vascko** [comment]
That sounds great, thanks!
I presume I'll need to compile the cri-o 1.19 branch manually since there is no RPMS for that version anywhere yet and the kubernetes minor version needs to match the cri-o minor version

**Comment by neolit123** [comment]
> anywhere yet and the kubernetes minor version needs to match the cri-o minor version

i wasn't aware of this fact. you can still try the 1,18 version of cri-o and see if is compatible with the above CI build that i linked.

**Comment by Vascko** [comment]
It looks like the process is getting a little further on v1.19-alpha.2. Kubelet still fails to start but this time with the following error message:
`Failed to start ContainerManager failed to get rootfs info: unable to find data in memory cache`

**Comment by neolit123** [comment]
ping @giuseppe @odinuge 
hi, do you happen to know what might be causing that?

**Comment by tedyu** [comment]
```
        if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LocalStorageCapacityIsolation) {
                rootfs, err := cm.cadvisorInterface.RootFsInfo()
                if err != nil {
                        return fmt.Errorf("failed to get rootfs info: %v", err)
```
The error was reported when LocalStorageCapacityIsolation is enabled.

For cadvisor, the 'unable to find data' seems to be a warning.

**Comment by Vascko** [comment]
Okay, disabled `--feature-gates="LocalStorageCapacityIsolation=false"` (what's the story behind the default value causing this error?)

Next error that it hits is this one:
`2069 qos_container_manager_linux.go:138] [ContainerManager] Failed to reserve QoS requests: failed to set resources for cgroup [kubepods burstable]: error while setting cgroup v2: [libcontainer AllowAllDevices is not supported, use Devices`

**Comment by tedyu** [comment]
The above error was fixed in runc:
```
commit 75a92ea615c81ae837165a42f553167a1124dede
Author: Kir Kolyshkin <kolyshkin@gmail.com>
Date:   Wed Apr 22 11:05:10 2020 -0700

    cgroupv2: allow to set EnableAllDevices=true
```
But it seems there hasn't been a new release since Apr 22nd.

**Comment by neolit123** [comment]
^ @kubernetes/sig-node-bugs 
for visibility on the above issues.

**Comment by giuseppe** [comment]
yes, there are still some open issues in cAdvisor, I am addressing them here: https://github.com/google/cadvisor/pull/2469

The cAdvisor PR is currently blocked on a new runc release, I hope we can move forward without it

**Comment by Vascko** [comment]
I see the cadvisor PR has been merged. Is that going to be in v1.19? Couldn't find it in the commits for v1.19

**Comment by Vascko** [comment]
Looks like runc still hassn't had a release after rc10 (rc90 seems to just fix a versioning issue)

**Comment by aaraodeo** [comment]
With kubelet 1.8.6 able to start a k8s cluster with cgroupfs v1(systemd.unified_cgroup_hierarchy = 0) on Fedora 31 server.
But when i upgraded kubelet to 1.19.0. i am not able to start k8s cluster. Its giving the same error described in the first update of the issue. I am using compatible version of kubeadm to setup the single master k8s cluster.

I0503 19:08:32.558943   11492 waitcontrolplane.go:87] [wait-control-plane] Waiting for the API server to be healthy
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
[kubelet-check] Initial timeout of 40s passed.
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get http://localhost:10248/healthz: dial tcp [::1]:10248: connect: connection refused.

Is cgroupfs v1 not supported with kubelet 1.19.0 on wards? Does kubelet 1.19.0 provides full support for cgroupfs v2 or is it still under development?

**Comment by danberindei** [comment]
/remove-lifecycle stale

I get the same error on Fedora 32 with kind v0.9.0, kubectl v1.20.0.

I tried images kindest/node:v1.19.1 (the default) and kindest/node:v1.18.8.

```
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get http://localhost:10248/healthz: dial tcp [::1]:10248: connect: connection refused.

	Unfortunately, an error has occurred:
		timed out waiting for the condition

	This error is likely caused by:
		- The kubelet is not running
		- The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)

	If you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:
		- 'systemctl status kubelet'
		- 'journalctl -xeu kubelet'

	Additionally, a control plane component may have crashed or exited when started by the container runtime.
	To troubleshoot, list all containers using your preferred container runtimes CLI.

	Here is one example how you may list all Kubernetes containers running in cri-o/containerd using crictl:
		- 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock ps -a | grep kube | grep -v pause'
		Once you have found the failing container, you can inspect its logs with:
		- 'crictl --runtime-endpoint unix:///run/containerd/containerd.sock logs CONTAINERID'
```

I'm not sure if I'm supposed to run those commands on my machine or in the kind container, but I have no `kubelet` SystemD unit and by the time this error message appears I don't have any running Docker container either.

**Comment by BenTheElder** [comment]
Kind runs in cgroupsv2 at HEAD, not sure about older versions. We have fedora CI covering this in our repo.

**Comment by BenTheElder** [comment]
> Is cgroupfs v1 not supported with kubelet 1.19.0 on wards? 

No it's supported and nearly all Kubernetes CI is cgroup v1 to this day.

> Does kubelet 1.19.0 provides full support for cgroupfs v2 or is it still under development?

It provides support. It doesn't support _rootless_ yet but it supports cgroupsv2.

(btw the release notes would cover both of these, dropping v1 would be a major breaking change, and v2 is mentioned in the 1.19 notes)

https://github.com/kubernetes/kubernetes/issues/90710#issuecomment-646336920

There have since been a few more ​runc releases and it's about to release 1.0.0

I think everything in this issue has been patched for some time now. Please file a new issue for any of these that remain (this has diverged a bit into a few only partially related issues, which should all be fixed).


Regarding kind specifically:
- use the latest release
- ensure you're using Kubernetes >= 1.19 https://kind.sigs.k8s.io/docs/user/known-issues/#failure-to-create-cluster-with-cgroups-v2
- if rootless, follow the rootless guide https://kind.sigs.k8s.io/docs/user/rootless/
